app/services/variant_deleter.rb

```diff
@@ -0,0 +1,17 @@
+# Checks the validity of a soft-delete call.
+class VariantDeleter
+  def delete(variant)
+    if only_variant_on_product?(variant)
+      variant.errors.add :product, I18n.t(:spree_variant_product_error)
+      return false
+    end
+
+    variant.destroy
+  end
+
+  private
+
+  def only_variant_on_product?(variant)
+    variant.product.variants == [variant]
+  end
+end
```

Comment:
Are we sure Spree doesn't deal with this case already? In any case, it feels to me this fits better in a model validation, which I suspect is already implemented by Spree.

Reply:
It is not implemented by Spree. In Spree, variants are optional. If a product doesn't have variants, it always has the hidden master variant holding the data of the product.

In OFN, we always have variants and we ignore the master variant. But if the model enforces that a product needs at least one variant, we can't delete a product, because we can't delete the last variant. We would need to delete without callbacks, but that could miss other important callbacks.

If there was a good way of implementing it on the model, I would go for it. Do you have any ideas? One option I was thinking of is `validates variants, presence: true` on the product. But the product is Spree code and that would break a lot of other cases. That's why I thought it's better to implement this outside of Spree and leave Spree code work as it is.

Reply:
Do you know why we chose to ignore the master and always have a variant?

My ideas was to move that logic you implemented to a custom validation method in the variant decorator.

Also, I restriction like this one should be constrainted at the DB level as well, but I don't see how now.

Reply:
> Do you know why we chose to ignore the master and always have a variant?

I'm not sure. I just remember that it was confusing to have normal variants and a master variant that is only active if there is no other variant. I think earlier in the project they decided to hide the master variant and deal only with real variants.

> My ideas was to move that logic you implemented to a custom validation method in the variant decorator.

Do you know if a validation on variants can distinguish if the destroy is only for the current variant or if the product is going to be destroyed as well? If we can't distinguish between these cases, we can't implement it on the model. The destroy is valid when the product is destroyed, but the destroy is invalid if the product persists and the last variant is to be destroyed.

Reply:
> The destroy is valid when the product is destroyed, but the destroy is invalid if the product persists and the last variant is to be destroyed.

can't we simply check these conditions in the validation method? Did you find any problem by calling `product.variants == [variant]` from a `before_destroy`?

Reply:
There has to be one. We have to think harde.r

Reply:
One option... In a `before_destroy` callback of `Spree::Product`, we could set an instance variable in `Spree::Variant` to true (e.g. `@destroy_through_product` or `@skip_last_variant_validation`). Then, in `Spree::Variant`, we could skip the last variant validation if this is true.

There's [`marked_for_destruction?`](https://apidock.com/rails/v4.2.7/ActiveRecord/AutosaveAssociation/marked_for_destruction%3F) and [`destroyed_by_association`](https://apidock.com/rails/ActiveRecord/AutosaveAssociation/destroyed_by_association). but I don't think they're applicable here.

Reply:
Good research @kristinalim. I just looked into those two methods and they don't seem to help. Your suggestion about the internal state variable is the only other option I can think of. So we have these two options:

* We add a `Product.before_destroy` callback to set `@requires_variants = false` and check that variable during `Variant.before_destroy`.
* We check for the validity of `Variant.destroy` in controller actions.

Since we are experiencing callback hell already I find the second option much more attractive. @luisramos0 @kristinalim What do you think? Or should we move this into a separate issue?

Reply:
Let's see what @kristinalim thinks. I have the feeling that this tightly related to the Variant model and might be legit to have callbacks for it but it's also right that we're experiencing the callback hell. If you want to give this approach a try, fine for me :+1:

Reply:
I think ofn and rails in general have one major problem which is models that are bloated and include business logic. the solution for this is to take as much as possible out of the models even if it is data manipulation like in this case. it helps keep the models simple and clean: a difficult thing to do!
this means I think model callbacks should be forbidden in this project at this scale :-D
Every action you do on a model should be explicit either in the model itself or in a service layer as in this case. variant_deleter is a very nice service and we should have more of them.

also, we have a working solution here so I think we should merge this PR and move on to different challenges ;-)

Reply:
Presence of at least one non-master variant is an assumption that we use in many places. Enforcing this in the model level is safer, and will help ensure that product data will always be valid.

Reply:
I think it's worth having a model callback for. :+1:

Reply:
Oh sorry, @luisramos0, I didn't see your comment.

If that's the direction that OFN is taking, not just for `Spree::Variant.destroy`, then I'm okay with this.

Reply:
Okay, so we have two proposals and half of us agree on one and the other half on the other. I find both solutions acceptable. One will make it harder to maintain the code, the other risks invalid data.

![Dilemma](https://media.giphy.com/media/xThuWmKYZKsZJG7I1a/giphy-downsized.gif)

Well, my argument would be that a product without variants is still valid. We just want to prevent users from deleting the last variant. Nothing should blow up if we had such a product in the database.

I'm actually a big fan of creating robust code that doesn't rely on many constraints. That makes it more flexible and stable. Imagine we wanted to replace a variant. Wouldn't it be nice if we could remove the current variant and add another one without it complaining it between?

Or imagine we decided to allow deleting the last variant and when that happens we just check for an empty product and delete that one as well. And that could be done in the VariantDeleter without changing or removing any callbacks and validations. Life could be that simple! :-D

![Simple life](https://media.giphy.com/media/12LNmxmmF47LIQ/giphy-downsized.gif)

Reply:
ehehe, interesting.
@kristinalim "If that's the direction that OFN is taking" it's only my opinion :-D

Reply:
Just to give some context to my opinion: it comes from knowing other systems that use datamapper pattern ORM, not AR ORM like Rails, and these large scale systems are more complex _and_ look much **better than all this callback stuff that is already unmanageable at this small scale**. The better solution is to take this logic out of models into services and **repositories** that handle logic and data in separate place from the models and their callbacks themselves.

Reply:
Agree on trying that @kristinalim ?

Reply:
Trying which, the `VariantDeleter` option? Yes, I'm okay with it. :slightly_smiling_face: Sounds good!

Reply:
Great. :+1: